### ISSUE TITLE
Elasticsearch Documentation

### ISSUE THREAD

**Issue by drphero** [Yakabuff/redarc#16]
Can we get some more documentation on elasticsearch? docker-compose-es.yml seems to get it running, but setting the host to http://localhost doesn't seem to work as I get `"curl: (7) Failed to connect to localhost port 9200 after 2 ms: Connection refused" when trying to use es_batch.sh`. I noticed that the docker container created by compose doesn't mention port 9200 anywhere so I tried adding them in, but get the same error.

EDIT:
Changing the ports in compose to "127.0.0.1:9200:9200" seems to get the connection working. But now I'm getting this error,
```
split: illegal option -- -
usage: split [-l line_count] [-a suffix_length] [file [prefix]]
       split -b byte_count[K|k|M|m|G|g] [-a suffix_length] [file [prefix]]
       split -n chunk_count [-a suffix_length] [file [prefix]]
       split -p pattern [-a suffix_length] [file [prefix]]
split: illegal option -- -
usage: split [-l line_count] [-a suffix_length] [file [prefix]]
       split -b byte_count[K|k|M|m|G|g] [-a suffix_length] [file [prefix]]
       split -n chunk_count [-a suffix_length] [file [prefix]]
       split -p pattern [-a suffix_length] [file [prefix]]
Warning: Couldn't read data from file "./torrents_es_sub.*", this makes an
Warning: empty POST.
{"error":{"root_cause":[{"type":"parse_exception","reason":"request body is required"}],"type":"parse_exception","reason":"request body is required"},"status":400}Warning: Couldn't read data from file "./torrents_es_com.*", this makes an
Warning: empty POST.
{"error":{"root_cause":[{"type":"parse_exception","reason":"request body is required"}],"type":"parse_exception","reason":"request body is required"},"status":400}%
```

EDIT 2:
I had to remove --verbose from the split commands in es_batch.sh. It seems to run fine, but the search still doesn't work. I get:  Error 500. Something went wrong or searching is disabled 

EDIT 3:
The data appears to be in the volume in the es01 container:
![Untitled](https://github.com/Yakabuff/redarc/assets/14174687/32e6bf41-479e-4e0a-b69c-361bab61e7a6)

However, the file size of 19.2MB matches the processed submission size, but 86.9MB does not match the match the processed comments size, which is 94.1MB. I'm not sure if this is the problem.

**Comment by Yakabuff** [comment]
Can you try querying elasticsearch index stats with CURL?

```
curl -u elastic:<password_here> -X GET "localhost:9200/redarc/_stats?pretty"

```

How are you running `es_batch.sh`?

**Comment by Yakabuff** [comment]
I've tested the `split` command it works fine with `--verbose`.  What OS and version of `split` are you using?

**Comment by drphero** [comment]
> I've tested the `split` command it works fine with `--verbose`. What OS and version of `split` are you using?

macOS 12.6.6
I'm not sure how to get the version of split.

**Comment by Yakabuff** [comment]
Hmm I'm using Ubuntu 22.04 with GNU coreutils

`split --version`

**Comment by drphero** [comment]
> Hmm I'm using Ubuntu 22.04 with GNU coreutils
> 
> `split --version`

Doesn't work because according to the man pages for split on mac, there is no --version or anything like it.

**Comment by Yakabuff** [comment]
Well as long as it creates `torrents_es_sub.*` files, it should be fine.

Can you try querying elasticsearch with curl? 

`curl -u elastic:<password_here> -X GET "localhost:9200/redarc/_stats?pretty"`

and if that works, a simple text search?
```
curl -u elastic:<password_here> -X GET "localhost:9200/redarc/_search?pretty" -H 'Content-Type: application/json' -d'
{
  "query": {
    "match": {
      "self_text": "torrent"
    }
  }
}
' 
```

**Comment by drphero** [comment]
> 



> Can you try querying elasticsearch index stats with CURL?
> 
> ```
> curl -u elastic:<password_here> -X GET "localhost:9200/redarc/_stats?pretty"
> ```
> 
> How are you running `es_batch.sh`?

I'm using './scripts/es_batch.sh torrents torrents_submissions torrents_comments estest1234'

curl response:

{
  "_shards" : {
    "total" : 2,
    "successful" : 1,
    "failed" : 0
  },
  "_all" : {
    "primaries" : {
      "docs" : {
        "count" : 52325,
        "deleted" : 0
      },
      "shard_stats" : {
        "total_count" : 1
      },
      "store" : {
        "size_in_bytes" : 20028986,
        "total_data_set_size_in_bytes" : 20028986,
        "reserved_in_bytes" : 0
      },
      "indexing" : {
        "index_total" : 52325,
        "index_time_in_millis" : 2511,
        "index_current" : 0,
        "index_failed" : 0,
        "delete_total" : 0,
        "delete_time_in_millis" : 0,
        "delete_current" : 0,
        "noop_update_total" : 0,
        "is_throttled" : false,
        "throttle_time_in_millis" : 0,
        "write_load" : 7.361347581921273E-4
      },
      "get" : {
        "total" : 0,
        "time_in_millis" : 0,
        "exists_total" : 0,
        "exists_time_in_millis" : 0,
        "missing_total" : 0,
        "missing_time_in_millis" : 0,
        "current" : 0
      },
      "search" : {
        "open_contexts" : 0,
        "query_total" : 0,
        "query_time_in_millis" : 0,
        "query_current" : 0,
        "fetch_total" : 0,
        "fetch_time_in_millis" : 0,
        "fetch_current" : 0,
        "scroll_total" : 0,
        "scroll_time_in_millis" : 0,
        "scroll_current" : 0,
        "suggest_total" : 0,
        "suggest_time_in_millis" : 0,
        "suggest_current" : 0
      },
      "merges" : {
        "current" : 0,
        "current_docs" : 0,
        "current_size_in_bytes" : 0,
        "total" : 0,
        "total_time_in_millis" : 0,
        "total_docs" : 0,
        "total_size_in_bytes" : 0,
        "total_stopped_time_in_millis" : 0,
        "total_throttled_time_in_millis" : 0,
        "total_auto_throttle_in_bytes" : 20971520
      },
      "refresh" : {
        "total" : 5,
        "total_time_in_millis" : 3,
        "external_total" : 3,
        "external_total_time_in_millis" : 159,
        "listeners" : 0
      },
      "flush" : {
        "total" : 2,
        "periodic" : 2,
        "total_time_in_millis" : 2042
      },
      "warmer" : {
        "current" : 0,
        "total" : 2,
        "total_time_in_millis" : 9
      },
      "query_cache" : {
        "memory_size_in_bytes" : 0,
        "total_count" : 0,
        "hit_count" : 0,
        "miss_count" : 0,
        "cache_size" : 0,
        "cache_count" : 0,
        "evictions" : 0
      },
      "fielddata" : {
        "memory_size_in_bytes" : 0,
        "evictions" : 0,
        "global_ordinals" : {
          "build_time_in_millis" : 0
        }
      },
      "completion" : {
        "size_in_bytes" : 0
      },
      "segments" : {
        "count" : 1,
        "memory_in_bytes" : 0,
        "terms_memory_in_bytes" : 0,
        "stored_fields_memory_in_bytes" : 0,
        "term_vectors_memory_in_bytes" : 0,
        "norms_memory_in_bytes" : 0,
        "points_memory_in_bytes" : 0,
        "doc_values_memory_in_bytes" : 0,
        "index_writer_memory_in_bytes" : 0,
        "version_map_memory_in_bytes" : 0,
        "fixed_bit_set_memory_in_bytes" : 0,
        "max_unsafe_auto_id_timestamp" : -1,
        "file_sizes" : { }
      },
      "translog" : {
        "operations" : 0,
        "size_in_bytes" : 55,
        "uncommitted_operations" : 0,
        "uncommitted_size_in_bytes" : 55,
        "earliest_last_modified_age" : 3215851
      },
      "request_cache" : {
        "memory_size_in_bytes" : 0,
        "evictions" : 0,
        "hit_count" : 0,
        "miss_count" : 0
      },
      "recovery" : {
        "current_as_source" : 0,
        "current_as_target" : 0,
        "throttle_time_in_millis" : 0
      },
      "bulk" : {
        "total_operations" : 1,
        "total_time_in_millis" : 2707,
        "total_size_in_bytes" : 16602384,
        "avg_time_in_millis" : 270,
        "avg_size_in_bytes" : 1660238
      }
    },
    "total" : {
      "docs" : {
        "count" : 52325,
        "deleted" : 0
      },
      "shard_stats" : {
        "total_count" : 1
      },
      "store" : {
        "size_in_bytes" : 20028986,
        "total_data_set_size_in_bytes" : 20028986,
        "reserved_in_bytes" : 0
      },
      "indexing" : {
        "index_total" : 52325,
        "index_time_in_millis" : 2511,
        "index_current" : 0,
        "index_failed" : 0,
        "delete_total" : 0,
        "delete_time_in_millis" : 0,
        "delete_current" : 0,
        "noop_update_total" : 0,
        "is_throttled" : false,
        "throttle_time_in_millis" : 0,
        "write_load" : 7.361347581921273E-4
      },
      "get" : {
        "total" : 0,
        "time_in_millis" : 0,
        "exists_total" : 0,
        "exists_time_in_millis" : 0,
        "missing_total" : 0,
        "missing_time_in_millis" : 0,
        "current" : 0
      },
      "search" : {
        "open_contexts" : 0,
        "query_total" : 0,
        "query_time_in_millis" : 0,
        "query_current" : 0,
        "fetch_total" : 0,
        "fetch_time_in_millis" : 0,
        "fetch_current" : 0,
        "scroll_total" : 0,
        "scroll_time_in_millis" : 0,
        "scroll_current" : 0,
        "suggest_total" : 0,
        "suggest_time_in_millis" : 0,
        "suggest_current" : 0
      },
      "merges" : {
        "current" : 0,
        "current_docs" : 0,
        "current_size_in_bytes" : 0,
        "total" : 0,
        "total_time_in_millis" : 0,
        "total_docs" : 0,
        "total_size_in_bytes" : 0,
        "total_stopped_time_in_millis" : 0,
        "total_throttled_time_in_millis" : 0,
        "total_auto_throttle_in_bytes" : 20971520
      },
      "refresh" : {
        "total" : 5,
        "total_time_in_millis" : 3,
        "external_total" : 3,
        "external_total_time_in_millis" : 159,
        "listeners" : 0
      },
      "flush" : {
        "total" : 2,
        "periodic" : 2,
        "total_time_in_millis" : 2042
      },
      "warmer" : {
        "current" : 0,
        "total" : 2,
        "total_time_in_millis" : 9
      },
      "query_cache" : {
        "memory_size_in_bytes" : 0,
        "total_count" : 0,
        "hit_count" : 0,
        "miss_count" : 0,
        "cache_size" : 0,
        "cache_count" : 0,
        "evictions" : 0
      },
      "fielddata" : {
        "memory_size_in_bytes" : 0,
        "evictions" : 0,
        "global_ordinals" : {
          "build_time_in_millis" : 0
        }
      },
      "completion" : {
        "size_in_bytes" : 0
      },
      "segments" : {
        "count" : 1,
        "memory_in_bytes" : 0,
        "terms_memory_in_bytes" : 0,
        "stored_fields_memory_in_bytes" : 0,
        "term_vectors_memory_in_bytes" : 0,
        "norms_memory_in_bytes" : 0,
        "points_memory_in_bytes" : 0,
        "doc_values_memory_in_bytes" : 0,
        "index_writer_memory_in_bytes" : 0,
        "version_map_memory_in_bytes" : 0,
        "fixed_bit_set_memory_in_bytes" : 0,
        "max_unsafe_auto_id_timestamp" : -1,
        "file_sizes" : { }
      },
      "translog" : {
        "operations" : 0,
        "size_in_bytes" : 55,
        "uncommitted_operations" : 0,
        "uncommitted_size_in_bytes" : 55,
        "earliest_last_modified_age" : 3215851
      },
      "request_cache" : {
        "memory_size_in_bytes" : 0,
        "evictions" : 0,
        "hit_count" : 0,
        "miss_count" : 0
      },
      "recovery" : {
        "current_as_source" : 0,
        "current_as_target" : 0,
        "throttle_time_in_millis" : 0
      },
      "bulk" : {
        "total_operations" : 1,
        "total_time_in_millis" : 2707,
        "total_size_in_bytes" : 16602384,
        "avg_time_in_millis" : 270,
        "avg_size_in_bytes" : 1660238
      }
    }
  },
  "indices" : {
    "redarc" : {
      "uuid" : "Hamk2ZE8QeKSKG2Op_Eb5w",
      "health" : "yellow",
      "status" : "open",
      "primaries" : {
        "docs" : {
          "count" : 52325,
          "deleted" : 0
        },
        "shard_stats" : {
          "total_count" : 1
        },
        "store" : {
          "size_in_bytes" : 20028986,
          "total_data_set_size_in_bytes" : 20028986,
          "reserved_in_bytes" : 0
        },
        "indexing" : {
          "index_total" : 52325,
          "index_time_in_millis" : 2511,
          "index_current" : 0,
          "index_failed" : 0,
          "delete_total" : 0,
          "delete_time_in_millis" : 0,
          "delete_current" : 0,
          "noop_update_total" : 0,
          "is_throttled" : false,
          "throttle_time_in_millis" : 0,
          "write_load" : 7.361347581921273E-4
        },
        "get" : {
          "total" : 0,
          "time_in_millis" : 0,
          "exists_total" : 0,
          "exists_time_in_millis" : 0,
          "missing_total" : 0,
          "missing_time_in_millis" : 0,
          "current" : 0
        },
        "search" : {
          "open_contexts" : 0,
          "query_total" : 0,
          "query_time_in_millis" : 0,
          "query_current" : 0,
          "fetch_total" : 0,
          "fetch_time_in_millis" : 0,
          "fetch_current" : 0,
          "scroll_total" : 0,
          "scroll_time_in_millis" : 0,
          "scroll_current" : 0,
          "suggest_total" : 0,
          "suggest_time_in_millis" : 0,
          "suggest_current" : 0
        },
        "merges" : {
          "current" : 0,
          "current_docs" : 0,
          "current_size_in_bytes" : 0,
          "total" : 0,
          "total_time_in_millis" : 0,
          "total_docs" : 0,
          "total_size_in_bytes" : 0,
          "total_stopped_time_in_millis" : 0,
          "total_throttled_time_in_millis" : 0,
          "total_auto_throttle_in_bytes" : 20971520
        },
        "refresh" : {
          "total" : 5,
          "total_time_in_millis" : 3,
          "external_total" : 3,
          "external_total_time_in_millis" : 159,
          "listeners" : 0
        },
        "flush" : {
          "total" : 2,
          "periodic" : 2,
          "total_time_in_millis" : 2042
        },
        "warmer" : {
          "current" : 0,
          "total" : 2,
          "total_time_in_millis" : 9
        },
        "query_cache" : {
          "memory_size_in_bytes" : 0,
          "total_count" : 0,
          "hit_count" : 0,
          "miss_count" : 0,
          "cache_size" : 0,
          "cache_count" : 0,
          "evictions" : 0
        },
        "fielddata" : {
          "memory_size_in_bytes" : 0,
          "evictions" : 0,
          "global_ordinals" : {
            "build_time_in_millis" : 0
          }
        },
        "completion" : {
          "size_in_bytes" : 0
        },
        "segments" : {
          "count" : 1,
          "memory_in_bytes" : 0,
          "terms_memory_in_bytes" : 0,
          "stored_fields_memory_in_bytes" : 0,
          "term_vectors_memory_in_bytes" : 0,
          "norms_memory_in_bytes" : 0,
          "points_memory_in_bytes" : 0,
          "doc_values_memory_in_bytes" : 0,
          "index_writer_memory_in_bytes" : 0,
          "version_map_memory_in_bytes" : 0,
          "fixed_bit_set_memory_in_bytes" : 0,
          "max_unsafe_auto_id_timestamp" : -1,
          "file_sizes" : { }
        },
        "translog" : {
          "operations" : 0,
          "size_in_bytes" : 55,
          "uncommitted_operations" : 0,
          "uncommitted_size_in_bytes" : 55,
          "earliest_last_modified_age" : 3215851
        },
        "request_cache" : {
          "memory_size_in_bytes" : 0,
          "evictions" : 0,
          "hit_count" : 0,
          "miss_count" : 0
        },
        "recovery" : {
          "current_as_source" : 0,
          "current_as_target" : 0,
          "throttle_time_in_millis" : 0
        },
        "bulk" : {
          "total_operations" : 1,
          "total_time_in_millis" : 2707,
          "total_size_in_bytes" : 16602384,
          "avg_time_in_millis" : 270,
          "avg_size_in_bytes" : 1660238
        }
      },
      "total" : {
        "docs" : {
          "count" : 52325,
          "deleted" : 0
        },
        "shard_stats" : {
          "total_count" : 1
        },
        "store" : {
          "size_in_bytes" : 20028986,
          "total_data_set_size_in_bytes" : 20028986,
          "reserved_in_bytes" : 0
        },
        "indexing" : {
          "index_total" : 52325,
          "index_time_in_millis" : 2511,
          "index_current" : 0,
          "index_failed" : 0,
          "delete_total" : 0,
          "delete_time_in_millis" : 0,
          "delete_current" : 0,
          "noop_update_total" : 0,
          "is_throttled" : false,
          "throttle_time_in_millis" : 0,
          "write_load" : 7.361347581921273E-4
        },
        "get" : {
          "total" : 0,
          "time_in_millis" : 0,
          "exists_total" : 0,
          "exists_time_in_millis" : 0,
          "missing_total" : 0,
          "missing_time_in_millis" : 0,
          "current" : 0
        },
        "search" : {
          "open_contexts" : 0,
          "query_total" : 0,
          "query_time_in_millis" : 0,
          "query_current" : 0,
          "fetch_total" : 0,
          "fetch_time_in_millis" : 0,
          "fetch_current" : 0,
          "scroll_total" : 0,
          "scroll_time_in_millis" : 0,
          "scroll_current" : 0,
          "suggest_total" : 0,
          "suggest_time_in_millis" : 0,
          "suggest_current" : 0
        },
        "merges" : {
          "current" : 0,
          "current_docs" : 0,
          "current_size_in_bytes" : 0,
          "total" : 0,
          "total_time_in_millis" : 0,
          "total_docs" : 0,
          "total_size_in_bytes" : 0,
          "total_stopped_time_in_millis" : 0,
          "total_throttled_time_in_millis" : 0,
          "total_auto_throttle_in_bytes" : 20971520
        },
        "refresh" : {
          "total" : 5,
          "total_time_in_millis" : 3,
          "external_total" : 3,
          "external_total_time_in_millis" : 159,
          "listeners" : 0
        },
        "flush" : {
          "total" : 2,
          "periodic" : 2,
          "total_time_in_millis" : 2042
        },
        "warmer" : {
          "current" : 0,
          "total" : 2,
          "total_time_in_millis" : 9
        },
        "query_cache" : {
          "memory_size_in_bytes" : 0,
          "total_count" : 0,
          "hit_count" : 0,
          "miss_count" : 0,
          "cache_size" : 0,
          "cache_count" : 0,
          "evictions" : 0
        },
        "fielddata" : {
          "memory_size_in_bytes" : 0,
          "evictions" : 0,
          "global_ordinals" : {
            "build_time_in_millis" : 0
          }
        },
        "completion" : {
          "size_in_bytes" : 0
        },
        "segments" : {
          "count" : 1,
          "memory_in_bytes" : 0,
          "terms_memory_in_bytes" : 0,
          "stored_fields_memory_in_bytes" : 0,
          "term_vectors_memory_in_bytes" : 0,
          "norms_memory_in_bytes" : 0,
          "points_memory_in_bytes" : 0,
          "doc_values_memory_in_bytes" : 0,
          "index_writer_memory_in_bytes" : 0,
          "version_map_memory_in_bytes" : 0,
          "fixed_bit_set_memory_in_bytes" : 0,
          "max_unsafe_auto_id_timestamp" : -1,
          "file_sizes" : { }
        },
        "translog" : {
          "operations" : 0,
          "size_in_bytes" : 55,
          "uncommitted_operations" : 0,
          "uncommitted_size_in_bytes" : 55,
          "earliest_last_modified_age" : 3215851
        },
        "request_cache" : {
          "memory_size_in_bytes" : 0,
          "evictions" : 0,
          "hit_count" : 0,
          "miss_count" : 0
        },
        "recovery" : {
          "current_as_source" : 0,
          "current_as_target" : 0,
          "throttle_time_in_millis" : 0
        },
        "bulk" : {
          "total_operations" : 1,
          "total_time_in_millis" : 2707,
          "total_size_in_bytes" : 16602384,
          "avg_time_in_millis" : 270,
          "avg_size_in_bytes" : 1660238
        }
      }
    }
  }
}

**Comment by drphero** [comment]
> Well as long as it creates `torrents_es_sub.*` files, it should be fine.
> 
> Can you try querying elasticsearch with curl?
> 
> `curl -u elastic:<password_here> -X GET "localhost:9200/redarc/_stats?pretty"`
> 
> and if that works, a simple text search?
> 
> ```
> curl -u elastic:<password_here> -X GET "localhost:9200/redarc/_search?pretty" -H 'Content-Type: application/json' -d'
> {
>   "query": {
>     "match": {
>       "self_text": "torrent"
>     }
>   }
> }
> ' 
> ```

It looks like the query is working. It's responding with several results.

**Comment by Yakabuff** [comment]
What error do you get in the browser dev console?  Is it ERRCONNREFUSED or some sort of error?

**Comment by drphero** [comment]
Just one error. For example:

XHR GET
http://localhost/api/search?type=submission&search=deluge&subreddit=torrents
[HTTP/1.1 500 Internal Server Error 203ms]

**Comment by Yakabuff** [comment]
Since we've confirmed that you can make queries with CURL, it's probably because the API isn't able to connect to elasticsearch.  

> Changing the ports in compose to "127.0.0.1:9200:9200"

I'm not sure this is correct.  Shouldn't it be `9200:9200`?

**Comment by drphero** [comment]
> Since we've confirmed that you can make queries with CURL, it's probably because the API isn't able to connect to elasticsearch.
> 
> > Changing the ports in compose to "127.0.0.1:9200:9200"
> 
> I'm not sure this is correct. Shouldn't it be `9200:9200`?

I had originally tried 9200:9200 several times with no luck.  The port wasn't getting bound. Now trying it again today, the port is now getting bound, but I am unable to communicate with the es01 container through port 9200. I receive `curl: (52) Empty reply from server` when running es_batch.sh and am not able to go to localhost:9200 in a browser. Starting again with the ports as 127.0.0.1:9200:9200 allows for communication on port 9200, but it seems the redarc container can't reach it then.

**Comment by Yakabuff** [comment]
Ok did a bit of investigating and this seems to work

Change ports back to `9200:9200`

Change `ES_HOST` value to `http://es01:9200`

**Comment by drphero** [comment]
> 

Tried with 9200:9200 and http://es01:9200, but couldn't run es_batch.sh. After setting the ports 127.0.0.1:9200:9200, it works! Thank you!

I noticed one more hiccup in the setup process. The first time running docker compose from a fresh start the tables aren't actually added to the database and so the load scripts don't actually load anything. I always had to remove the containers and run docker compose again for the start.sh script to add the tables. Maybe it's trying to add the tables before the database is ready?
